### PR TITLE
[Gutenberg] Unsupported block fallback

### PR DESCRIPTION
**Pull request by etoledom**
Proof of concept Unsupported block fallback on native side.
This opens a web view with a new-post editor, and insert the required block via "User Content Javascript" (JS code inserted to the web view).

We extract the final block content in the same way.

To clean up the canvas from unwanted UI elements we inject CSS that hides those elements. 

`gutenberg-mobile` side PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2063

To test:

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Gutenberg/GutenbergWeb/GutenbergWebNavigationViewController.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    /// Due to a bug on iOS 13, presenting a DocumentController on a modally presented controller will result on a crash.
+    /// This is a workaround to prevent this crash.
+    /// More info: https://stackoverflow.com/questions/58164583/wkwebview-with-the-new-ios13-modal-crash-when-a-file-picker-is-invoked
+    ///
+    override func present(_ viewControllerToPresent: UIViewController, animated flag: Bool, completion: (() -> Void)? = nil) {
+        if #available(iOS 13, *),
+            let menuViewControllerClass = NSClassFromString("UIDocumentMenuViewController"), // Silence deprecation warning.
+            viewControllerToPresent.isKind(of: menuViewControllerClass),
+            UIDevice.current.userInterfaceIdiom == .phone
+        {
```
[the remark this answers is not in the document]
Comment (etoledom) [reply WordPress/Classes/ViewRelated/Gutenberg/GutenbergWeb/GutenbergWebNavigationViewController.swift]: But this is ugly in this case :(

**Review by guarani — COMMENTED** [review]
This is fantastic @etoledom! 🎉 I feel like we've opened up a whole new realm of possibilities with this new functionality to edit "unsupported" blocks.

#### Thoughts from a user's perspective

These might be out of scope for a proof-of-concept, but I wanted to share now that I've had a chance to give this a whirl.

- Could we add an activity indicator as the block is loading in the web view? 
- As a user, I might think that tapping on Save from the unsupported blocks screen is enough to save changes. However it's not until I tap Update in the post itself that the changes are truly saved. We could change Save to Continue (or similar) to avoid confusion.
- The bottom sheet that explains that a block is not currently supported might need some modifications. Users that read "web browser" may be surprised to find that tapping the button does not navigate out of the app to mobile Safari as expected, but shows what looks like just another screen from the app itself.
- Some blocks have nice, capitalized names – but others appear like "coblocks/pricing-table" which is less user friendly.

#### Test scenarios

I edited unsupported blocks on mobile and verified that they are updated correctly on web. I also tried posts with multiple unsupported blocks and made sure to test a few random blocks I'd never used before like Navigation, Logos & Badges, and Price Table.

#### Going forward

Should we whitelist the blocks that are available for editing on the unsupported blocks screen, or can we assume that all blocks, present and future, can be edited with this view?

**Review by guarani — APPROVED** [review]

**Comment by etoledom** [comment]
Thank you @Guarani , this is valuable feedback!
I feel like @iamthomasbishop can answer better than me. I'll follow his lead!

Here's my opinion:
> Could we add an activity indicator as the block is loading in the web view?

Sure! Maybe the same progress indicator from the Preview would be enough?

>  I might think that tapping on Save from the unsupported blocks screen is enough to save changes.

Good point. It's not clear if the save is done locally or synced with remote. We could see users feedback before modifying this. "Continue" might still be not clear enough what it does.

> Users that read "web browser" may be surprised to find that tapping the button does not navigate out of the app to mobile Safari as expected

Yeah, "browser" might be confusing for some people, but it might be more clear for most that it's not a "native element". I'd propose writing "web view" but even that might be too technical for many.

> Some blocks have nice, capitalized names – but others appear like "coblocks/pricing-table" which is less user friendly.

I think we have the nice naming of core blocks (and now maybe Jetpack, not sure), but anything that it's not bundled we have only that raw kind of name. We could try to manipulate it (remove anything before the `/`, capitalising, etc...) but I'm not sure I'm comfortable with that. At least not at this stage.

**Comment by iamthomasbishop** [comment]
> Could we add an activity indicator as the block is loading in the web view?

@guarani this is a good idea, and one I assumed we wouldn't have the ability to do. If we _do_ have the ability to do this relatively easy, we should do it. Do we have access to the "spinner" that the web often uses ([this one](https://cloudup.com/cxt0Km4qPdG))? If not, we can do something like @etoledom mentioned and use the progress indicator we use on Preview.

> As a user, I might think that tapping on Save from the unsupported blocks screen is enough to save changes. However it's not until I tap Update in the post itself that the changes are truly saved. We could change Save to Continue (or similar) to avoid confusion.

This is a good point as well, and I think either `Continue` or something like `Apply` would be better. I had asked if we could show an indicator of what happened because of this confusion, but this (and changing `X` to `Cancel` or `Discard` [as discussed on the master issue](https://github.com/wordpress-mobile/gutenberg-mobile/pull/2063#issuecomment-622106212)) would be clearer. The labels would become either `Cancel`/`Discard` and `Continue`/`Apply`.

> Users that read "web browser" may be surprised to find that tapping the button does not navigate out of the app to mobile Safari as expected, but shows what looks like just another screen from the app itself.

Good point, and a tricky one. What about something like "edit using the in-app browser" or "edit using the web editor"? I'm not sure if either of these are better, but I agree it's not perfect as it stands.

**Comment by guarani** [comment]
Thanks for the response @iamthomasbishop! 

> Do we have access to the "spinner" that the web often uses (this one)?

I'll leave this to @etoledom to respond.

> The labels would become either Cancel/Discard and Continue/Apply.

I agree that either of these two options, "Continue" or "Apply", would make it clearer to the user that this is still only an intermediate step on the way to actually saving the modified content. (Personally, Apply still sounds to me like something will be persisted or saved whereas Continue sounds like changes will not be saved yet, but I could be wrong.)

> What about something like "edit using the in-app browser" or "edit using the web editor"?

Backing up a bit, this project brings support for _editing_ blocks that were previously not editable on the mobile apps. I would say that what we provide now is best described as the ability to edit block _settings_ for unsupported blocks. Looking at the bottom sheet perhaps we could simply use "Edit block settings" as the button title.

| Old | New |
| -- | -- |
| <kbd></kbd> | <kbd></kbd> |

I propose we

**Comment by iamthomasbishop** [comment]
> I would say that what we provide now is best described as the ability to edit block _settings_ for unsupported blocks.

@guarani  I don’t think this is only block settings, is it? From what I can tell, we aren’t placing special restrictions (other than hiding a few menu items) on what the user can edit on the canvas, so I don’t see a reason to frame it in that way. 

For example, let’s say there is some sort of third-party text-based block that we don’t yet support out-of-the-box on mobile. If they use this function to open the block in the web view, they should be able to edit almost any aspect of that blocks contents, correct? Perhaps I am missing something?

**Comment by guarani** [comment]
> @guarani I don’t think this is only block settings, is it? From what I can tell, we aren’t placing special restrictions (other than hiding a few menu items) on what the user can edit on the canvas, so I don’t see a reason to frame it in that way.
> 
> For example, let’s say there is some sort of third-party text-based block that we don’t yet support out-of-the-box on mobile. If they use this function to open the block in the web view, they should be able to edit almost any aspect of that blocks contents, correct? Perhaps I am missing something?

You're right, it's more than just block settings – I wasn't thinking past blocks similar to the video block which just have settings and no text content.

**Comment by iamthomasbishop** [comment]
Ah ok, no worries — just wanted to make sure I wasn't missing something 😆

**Comment by etoledom** [comment]
@iamthomasbishop - @guarani 
> Do we have access to the "spinner" that the web often uses (this one)? If not, we can do something like @etoledom mentioned and use the progress indicator we use on Preview.

I have implemented the Preview's loading indicator, since this is a native component we don't have easy access to it.

**Comment by loremattei** [comment]
Hey! I'm moving this to 15.0 because 14.9 has been cut. If you want this to make it to 14.9, please feel free to ping me.

**Comment by guarani** [comment]
@etoledom I took the liberty of updating the milestone to 15.1 since 15.0 is about to be cut.

**Comment by etoledom** [comment]
Updated this branch with current develop and `gutenberg` pointing to `gutenberg-mobile:develop`.

@Guarani - would you mind doing a last smoke test?

- With metro running on `gutenberg-mobile:develop`:
  - Unsupported block feature should work as expected.
- Running Gutenberg from bundle (no metro running):
  - No option to open the web view on unsupported blocks should be visible.
  - Check that there are no errors or red screens.

Thanks!

**Comment by guarani** [comment]
@etoledom – I got this working, but at first I was getting the login screen on a .com site on the simulator. On my phone I got a blank screen and noticed this error in the Xcode console:
```
Please update the following components: withAnimatable(Component)
Error Domain=WKErrorDomain Code=4 "A JavaScript exception occurred" UserInfo={WKJavaScriptExceptionLineNumber=1, WKJavaScriptExceptionMessage=TypeError: undefined is not a function (near '...window.insertBlock...'), WKJavaScriptExceptionColumnNumber=19, WKJavaScriptExceptionSourceURL=undefined, NSLocalizedDescription=A JavaScript exception occurred}
```
I reinstalled and it started working on the same device, but the error was still shown on the console. This error didn't seem to stop changes from being made to the unsupported block.

**Comment by etoledom** [comment]
> Please update the following components: withAnimatable(Component)

This seems to be related to the bottom sheet in general, not to this feature in particular.

> Error Domain=WKErrorDomain Code=4 "A JavaScript exception occurred" UserInfo={WKJavaScriptExceptionLineNumber=1, WKJavaScriptExceptionMessage=TypeError: undefined is not a function (near '...window.insertBlock...'), WKJavaScriptExceptionColumnNumber=19, WKJavaScriptExceptionSourceURL=undefined, NSLocalizedDescription=A JavaScript exception occurred}

This happens usually on .com, since there's an attempt to insert the block on page load, but the editor is not ready yet. There's a second attempt later on that will succeed.

**Comment by etoledom** [comment]
Merging to bring the gutenberg bridge changes to WPiOS.
The new feature still is under dev-flag on `gutenberg-mobile` side.

**Pull request merged**